[ILM] make IndexLifecycleExplainResponse more resilient to null values

talevy opened the pull request:
There is a time window where the index has a policy name, but is not really
managed by ILM just yet since it doesn't have its step info set.

This commit introduces a more resilient parser to null step details in the explain response.

elasticmachine commented:
Pinging @elastic/es-core-features

talevy commented:
test this please

colings86 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/IndexLifecycleExplainResponse.java:
```diff
@@ -73,10 +73,10 @@
+        PARSER.declareStringOrNull(ConstructingObjectParser.optionalConstructorArg(), PHASE_FIELD);
+        PARSER.declareStringOrNull(ConstructingObjectParser.optionalConstructorArg(), ACTION_FIELD);
+        PARSER.declareStringOrNull(ConstructingObjectParser.optionalConstructorArg(), STEP_FIELD);
+        PARSER.declareStringOrNull(ConstructingObjectParser.optionalConstructorArg(), FAILED_STEP_FIELD);
```
We might want to have a smarter check here I think. We only expect the case where all the fields are `null`, we never expect only one of the fields to be null so I think we should add a check int he constructor that ensure if one of the phase, action, step, failed_step fields are null they all are and otherwise the all have a value throwing an exception if the check fails.

colings86 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/IndexLifecycleExplainResponse.java:
```diff
@@ -63,10 +63,10 @@
+        PARSER.declareStringOrNull(ConstructingObjectParser.optionalConstructorArg(), PHASE_FIELD);
+        PARSER.declareStringOrNull(ConstructingObjectParser.optionalConstructorArg(), ACTION_FIELD);
+        PARSER.declareStringOrNull(ConstructingObjectParser.optionalConstructorArg(), STEP_FIELD);
+        PARSER.declareStringOrNull(ConstructingObjectParser.optionalConstructorArg(), FAILED_STEP_FIELD);
```
Same comment as above

talevy commented:
thanks for the review @colings86. Since, I agree, the step details (phase,action,step + times) are never expected to be partially null, I added verification for this in the constructor.

I also reverted the allowance for phase/action/step to be null since we can simply not render it, to be consistent with the rendering of their respective times.

Another thing that I changed is the format of the client-side response. I found it difficult to maintain the two subtly-different objects, so I migrated the usage of `long` for the fields, to `Long` (just as it is on the server-side).

talevy commented on client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/IndexLifecycleExplainResponse.java:
```diff
@@ -95,36 +96,46 @@
             }
+            long numNull = Stream.of(phase, action, step, phaseTime, actionTime, stepTime).filter(Objects::isNull).count();
+            if (numNull > 0 && numNull < 6) {
+                throw new IllegalArgumentException("invalid step details in response [" +
```
mention that this is an error in the response object, and that there needs to be complete information. Also added a code comment.

thanks for the suggestion!
